[Vue] migrate siteselector directive and quick-access directive

diosmosis opened the pull request:
### Description:

**This PR is based off of https://github.com/matomo-org/matomo/pull/18225.**

This PR has a breaking change to createAngularAdapter() so other Vue PRs in progress may need to be modified (unless they are merged before this, then I'll fix the conflicts here).

Changes:
* Add typescript types for mousetrap library.
* deprecate autocomplete-matched as it's trivial to implement in siteselector and is not used anywhere else.
* in broadcast.propagateNewPage() add extra parameter for supplying the entire new URL to load. in this case, it does not build the URL using other parameters.
* add a MatomoUrl.updateUrl() method that reloads the entire URL (search + hash)
* make sure promise returned by AjaxHelper.fetch has an abort() method
* in createAngularAdapter() add vue instance to the event handler functions and postCreate(). this is so ngModel could be connected to v-model in Vue.
* migrated the siteselector directive.
* migrated the quick-access directive.
* migrated the siteselector-model service as the SitesStore. This is shared between quick-access and siteselector. The only state that's shared is the initial list of sites that is loaded once, unlike currently where the list of currently displayed sites is also shared between the two directives.
* in createAngularAdapter() allow specifying a `require:` field to the angularjs directive that is created.
* add the ability to transform values when binding from angularjs => vue (for example, converting values to booleans, as in `noclear` for siteselector)
* disable webpack asset size hints (since our assets are already bigger at the end than what is recommended by webpack).
* when changing sites, removed the `showAjaxLoading = false` parameter so MatomoUrl.updateUrl() will stay simple. Doesn't seem to have much of an effect IMO so I think it's ok.

![image](https://user-images.githubusercontent.com/125140/141254156-3359ed88-2ce3-4036-accd-6971954706b4.png)

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

diosmosis committed:
migrating RateFeature and ReviewLinks + adding AjaxHelper.fetch utility method (all untested)

diosmosis committed:
get ratefeature component to work, modify matomodialog component to use v-model, add event parameters to createAngularAdapter, allow translate to use variadic args or one string array + rebuild

diosmosis committed:
remove ratefeature angularjs files

diosmosis committed:
rebuild + make vue mapping property optional in createANgularJsAdapter

diosmosis committed:
migrate enrichedheadline and get to work

diosmosis committed:
fix test

diosmosis committed:
fix translate

diosmosis committed:
fix another translate issue & migrate contentblock directive

diosmosis committed:
fix anchor links, not including the "/" causes angularjs to fail (also on 4.x-dev)

diosmosis committed:
update expected screenshots

diosmosis committed:
fix ui test

diosmosis committed:
fix some test failures

diosmosis committed:
fix nested transclude issue

diosmosis committed:
remove content block files

diosmosis committed:
fix icon spacing that occurs due to angularjs inserting empty comments in between nodes while vue 3 does not

diosmosis committed:
update some screenshots

diosmosis committed:
update screenshot (actually fixes an alignment issue)

diosmosis committed:
update screenshot

diosmosis committed:
first pass at converting comparisons service/component

diosmosis committed:
get new code to build and load without error in the UI

diosmosis committed:
debugging

diosmosis committed:
getting basic functionaltiy to work

diosmosis committed:
Update _dataTable.twig

diosmosis committed:
fix UI test failure + URL encoding/angularjs issue causing back button to not work

diosmosis committed:
fix order of operations issue

diosmosis committed:
built vue files

diosmosis committed:
using ref in setup() is not needed to access this.$refs

diosmosis committed:
Convert comparisons service angularjs tests to comparison store typescript tests.

diosmosis committed:
migrate piwik-date-picker directive

diosmosis committed:
migrate date range picker component (changed invalid date in input handling to just reset back to the previous date since it was easier in vue to do that)

diosmosis committed:
migrate period-date-picker component (using composition api more when easier for migration)

diosmosis committed:
convert piwik-expand-onclick directive to vue directive

diosmosis committed:
migrate expand on hover directive to vue directive

diosmosis committed:
fix variable reference

diosmosis committed:
Merge branch 'vue-content-block' into vue-date-picker

diosmosis committed:
build

diosmosis committed:
Add materialize-css @types and migrate piwik-dropdown-menu.

diosmosis committed:
Merge branch 'vue-content-block' into vue-dropdown

diosmosis committed:
migrate focus-anywhere-but-here directive to vue directive

diosmosis committed:
migrate focus-if directive

diosmosis committed:
migrate menudropdown directive

diosmosis committed:
forgot to remove old files

diosmosis committed:
Merge branch 'vue-date-picker' into vue-dropdown

diosmosis committed:
Merge branch '4.x-dev' into vue-comparisons

diosmosis committed:
Merge branch 'vue-comparisons' into vue-date-picker

diosmosis committed:
built vue files

diosmosis committed:
Merge branch 'vue-date-picker' into vue-dropdown

diosmosis committed:
first pass at migrating notification directive, notification service and parts of UI.Notification to Vue

diosmosis committed:
Merge branch '4.x-dev' into vue-comparisons

diosmosis committed:
Merge branch 'vue-comparisons' of github.com:matomo-org/matomo into vue-comparisons

diosmosis committed:
rewrite URL handling to use computed properties in a URL store + do the same for other dependent data in the comparison store to allow vues to subscribe to the properties for changes to global state

diosmosis committed:
fix some tests

diosmosis committed:
some more fixes

diosmosis committed:
more fixes + disallow modifications to MatomoUrl state

diosmosis committed:
get angularjs unit tests to pass + fix a couple more issues

diosmosis committed:
another fix

diosmosis committed:
fix bad merge

diosmosis committed:
self review + fixes

diosmosis committed:
Merge branch '4.x-dev' into vue-comparisons

diosmosis committed:
remove old fix as it may not be needed anymore

diosmosis committed:
empty string is not a valid date + do not report invalid date exception just rethrow

diosmosis committed:
update screenshots and try to fix random failure

diosmosis committed:
use jquery $destroy event instead of scope one since the scope one is broadcasted

diosmosis committed:
Merge branch 'vue-jquery-destroy' into vue-comparisons

diosmosis committed:
Merge branch 'vue-comparisons' into vue-date-picker

diosmosis committed:
Merge branch 'vue-date-picker' into vue-dropdown

diosmosis committed:
Merge branch 'vue-dropdown' into vue-notifications

diosmosis committed:
Merge branch '4.x-dev' into vue-date-picker

diosmosis committed:
rangeChange event must be triggered once on mount

diosmosis committed:
initialize startDateText/endDateText correctly

diosmosis committed:
use jquery $destroy event instead of angularjs one

diosmosis committed:
built vue files

diosmosis committed:
Merge branch 'vue-date-picker' into vue-dropdown

diosmosis committed:
fix menudropdown.directive.js reference

diosmosis committed:
load vue in installation/updater & correctly make focusanywherebuthere stateful

diosmosis committed:
correctly implement stateful directives for ExpandOnClick/ExpandOnHover

diosmosis committed:
Merge branch 'vue-date-picker' into vue-dropdown

diosmosis committed:
less tweak (angularjs comment removal)

diosmosis committed:
fix submenu check

diosmosis committed:
quick type fix

diosmosis committed:
load vue in installation workflow

diosmosis committed:
add broadcast.js to Installation workflow + do not fail in pk_translate if no translations are loaded

diosmosis committed:
update expected screenshots (spacing of arrow changed because of angularjs comment no longer being there)

diosmosis committed:
Merge branch 'vue-dropdown' into vue-notifications

diosmosis committed:
start moving Notification class code to notifications store

diosmosis committed:
fix prop type

diosmosis committed:
progress on toast migration

diosmosis committed:
fix html escaping

diosmosis committed:
Merge branch '4.x-dev' into vue-dropdown

diosmosis committed:
built vue files

diosmosis committed:
get toast and other transitions to work + fix broken toast

diosmosis committed:
move all of notification.js to NotificationStore

diosmosis committed:
Merge branch 'vue-dropdown' into vue-notifications

diosmosis committed:
wait for angular to be initialized to post events to avoid loading race condition

diosmosis committed:
Merge branch 'vue-matomo-event-wait' into vue-notifications

diosmosis committed:
get scroll to notification to work + get initialization of notification groups to work

diosmosis committed:
correct unmount + remove notifications service file

diosmosis committed:
fix some test failures

sgiehl committed:
Merge branch '4.x-dev' into vue-dropdown

diosmosis committed:
re add accidentally removed (?) file

diosmosis committed:
remove no longer needed file

diosmosis committed:
Add CoreHome UMD in CoreUpdater/Installation.

diosmosis committed:
self review

diosmosis committed:
Merge branch 'vue-dropdown' into vue-notifications

diosmosis committed:
fix type + add default value

diosmosis committed:
remove file from JS list

diosmosis committed:
Merge branch 'vue-dropdown' into vue-notifications

diosmosis committed:
fix test

diosmosis committed:
fix UI tests

diosmosis committed:
Merge branch 'vue-dropdown' into vue-notifications

diosmosis committed:
set correct type in users manager notification and allow scope values to be transformed in createAngularAdapter

diosmosis committed:
Merge branch '4.x-dev' into vue-notifications

diosmosis committed:
start migrating siteselector

diosmosis committed:
small addition

diosmosis committed:
migrate rest of site selector code + make some breaking changes to function signatures in createAngularJsAdapter

diosmosis committed:
disable webpack asset size hints/warnings + get siteselector code to build

diosmosis committed:
fixing some bugs

diosmosis committed:
fix some more issues (allow specifing require in createAngularJsAdapter and make AjaxHelper promises abortable)

diosmosis committed:
get npm test to pass

diosmosis committed:
a couple more fixes

diosmosis committed:
remove existing files

diosmosis committed:
convert quick-access directive and use shared code/state with site selector

diosmosis committed:
remove site selector model

diosmosis committed:
fix more issues and get UI tests to pass for quickaccess

diosmosis committed:
remove debugging code / todo

diosmosis committed:
fix initial value

diosmosis committed:
Merge branch '4.x-dev' into vue-siteselector

diosmosis committed:
add back a $timeout()

diosmosis committed:
Merge branch '4.x-dev' into vue-notifications

diosmosis committed:
Merge branch 'vue-notifications' into vue-siteselector

diosmosis committed:
fixing tests, the post blur scope.$apply()s are apparently required for angularjs to function properly

diosmosis committed:
fixing more UI test failures

diosmosis committed:
rebuild

diosmosis committed:
fix vue build

diosmosis committed:
why were these deleted?

diosmosis committed:
remove debug code

diosmosis committed:
fix css class issue + update expected screenshots

diosmosis committed:
rebuild CoreHome

diosmosis commented:
Explanation for UsersManager screenshot updates:

The users manager root component gets set an initial site ID. Since there is not idSite in the URL in the test, it is set to 1. The site w/ ID = 1 is SLEEP, which is correct now. Before it seems it wasn't being initialized correctly in the site selector.

diosmosis committed:
Merge branch '4.x-dev' into vue-siteselector

diosmosis committed:
revert styling change

diosmosis committed:
Merge branch '4.x-dev' into vue-notifications

diosmosis committed:
built vue files

diosmosis committed:
get focus-if to work and remove debugging return;

diosmosis committed:
rebuilt vue

diosmosis committed:
Merge branch 'vue-notifications' into vue-siteselector

diosmosis committed:
should not need to specify type there

diosmosis committed:
built CoreHome

diosmosis committed:
Merge branch '4.x-dev' into vue-siteselector

diosmosis committed:
built vue files

peterhashair reviewed (commented):
That works for me perfectly, maybe still need to do resolve conflicts, also just a tiny suggestion maybe on key up autocomplete event maybe we need a debounce.

diosmosis commented on plugins/CoreHome/vue/src/QuickAccess/QuickAccess.vue:
```diff
@@ -0,0 +1,461 @@
+      } else if (isEscKey && areSearchResultsDisplayed) {
+        this.deactivateSearch();
+      } else {
+        setTimeout(() => {
```
There wasn't one before but we can add one.

peterhashair commented:
@diosmosis also I got an question for function naming. 
 like that one, do we need a camel case? 
```Javascript 
export { default as Menudropdown } from './Menudropdown/Menudropdown.vue';
```

diosmosis commented:
> I got an question for function naming. like that one, do we need a camel case?

Yes, that would probably be better :+1:

diosmosis committed:
apply review feedback

diosmosis commented:
converted to draft so it won't get merged until after the 4.6.0 release

justinvelluppillai commented:
> converted to draft so it won't get merged until after the 4.6.0 release

PRs merged to 4.x-dev won't be released in 4.6.0

diosmosis committed:
Merge branch '4.x-dev' into vue-siteselector

diosmosis commented:
@justinvelluppillai ok, removing draft status

peterhashair reviewed (approved):
looks good to me.

sgiehl commented:
@diosmosis I'm currently doing some local testing of the changes. Will leave some more comments when I find something else.

It seems the search results in site selector are not escaped correctly. I have a website with the name `<a href="#">?</a>`, it displays correctly when listed on opening the siteselector. But when searching, it is only displayed as `?`, the html tags seems to be inserted. So maybe a script tag might evaluate here as well.

diosmosis commented:
> So maybe a script tag might evaluate here as well.

It's sanitized so dangerous HTML is removed. The other html I guess isn't being escaped properly.

diosmosis committed:
get auto clearing behavior to work in site selector

diosmosis committed:
fix a couple more bugs

diosmosis commented:
> It seems the search results in site selector are not escaped correctly. I have a website with the name <a href="#">?</a>, it displays correctly when listed on opening the siteselector. But when searching, it is only displayed as ?, the html tags seems to be inserted. So maybe a script tag might evaluate here as well.

Can you check again? this is working for me:

![image](https://user-images.githubusercontent.com/125140/142291559-e6a7a9d6-df2c-42f6-b38e-875a9eaa35ac.png)

> When I type something into the search box in site selector and then click on the input field again, the field either looses it's content (if the request is still pending), or it sometimes even disappears (when the request had finished)

Looks like there's a bug in the original code. Should be fixed in the vue PR now.

> Pressing the f key highlights the quick search visually, but doesn't give the input box the focus anymore

Fixed.

> Pressing the w key opens and highlights the site selector visually, but doesn't give the input box the focus anymore

Fixed.

> Unrelated to the changes here. But pressing d doesn't open the date picker any longer. Maybe that's due to other vue changes before

Found the issue, will fix in PR to next_release.

diosmosis committed:
rebuild vue

sgiehl commented:
@diosmosis hm. funny. It works when searching for `?` for me as well. But when searching for any letter that actually isn't in the site's name, it doesn't.

![image](https://user-images.githubusercontent.com/1579355/142379858-c7266fb6-7645-4840-97e0-cac61ee3f83d.png)

diosmosis committed:
escape htmle entities in site name

diosmosis commented:
@sgiehl should be fixed. Looks like the issue only occurred when searching for html entity letters (eg, searching for `gt` which in the backend will match `&lt;img onsrc...`).

diosmosis committed:
tweak

sgiehl reviewed (approved):
works now 👍
